### PR TITLE
Add -s SUPPORT_POSIX_ENV=0 linker flag. Fixes #11148.

### LINKED ISSUE #11148: Unused POSIXism leaks into compiled apps (+3KB, ~+2.3% overhead)
In an application that does not use any of POSIX environment code ([setenv, getenv, environ, etc.](https://www.gnu.org/software/libc/manual/html_node/Environment-Access.html)), and does not use the MEMFS filesystem or WASI, the following code creeps into the build:

Wasm side:
```wast
 (import "a" "Eb" (func $__wasi_environ_sizes_get (param i32 i32) (result i32)))
 (import "a" "Fb" (func $__wasi_environ_get (param i32 i32) (result i32)))

 (func $__wasm_call_ctors
  (call $initialize_malloc_heap\28\29)
  (call $__emscripten_environ_constructor)
  ...
)

 (func $__emscripten_environ_constructor
  (local $0 i32)
  (local $1 i32)
  (local $2 i32)
  (global.set $global$0
   (local.tee $0
    (i32.sub
     (global.get $global$0)
     (i32.const 16)
    )
   )
  )
  (block $label$1
   (br_if $label$1
    (call $__wasi_environ_sizes_get
     (i32.add
      (local.get $0)
      (i32.const 12)
     )
     (i32.add
      (local.get $0)
      (i32.const 8)
     )
    )
   )
   (i32.store
    (i32.const 282472)
    (local.tee $1
     (call $emmalloc_memalign
      (i32.add
       (i32.shl
        (i32.load offset=12
         (local.get $0)
        )
        (i32.const 2)
       )
       (i32.const 4)
      )
     )
    )
   )
   (br_if $label$1
    (i32.eqz
     (local.get $1)
    )
   )
   (block $label$2
    (if
     (local.tee $1
      (call $emmalloc_memalign
       (i32.load offset=8
        (local.get $0)
       )
      )
     )
     (br_if $label$2
      (local.tee $2
       (i32.load
        (i32.const 282472)
       )
      )
     )
    )
    (i32.store
     (i32.const 282472)
     (i32.const 0)
    )
    (br $label$1)
   )
   (i32.store
    (i32.add
     (local.get $2)
     (i32.shl
      (i32.load offset=12
       (local.get $0)
      )
      (i32.const 2)
     )
    )
    (i32.const 0)
   )
   (br_if $label$1
    (i32.eqz
     (call $__wasi_environ_get
      (i32.load
       (i32.const 282472)
      )
      (local.get $1)
     )
    )
   )
   (i32.store
    (i32.const 282472)
    (i32.const 0)
   )
  )
  (global.set $global$0
   (i32.add
    (local.get $0)
    (i32.const 16)
   )
  )
 )
```

JS side:

```js
var ENV = {};

function __getExecutableName() {
    return "./this.program"
}

function getEnvStrings() {
    if (!getEnvStrings.strings) {
        var env = {
            "USER": "web_user",
            "LOGNAME": "web_user",
            "PATH": "/",
            "PWD": "/",
            "HOME": "/home/web_user",
            "LANG": (typeof navigator === "object" && navigator.languages && navigator.languages[0] || "C").replace("-", "_") + ".UTF-8",
            "_": __getExecutableName()
        };
        for (var x in ENV) {
            env[x] = ENV[x]
        }
        var strings = [];
        for (var x in env) {
            strings.push(x + "=" + env[x])
        }
        getEnvStrings.strings = strings
    }
    return getEnvStrings.strings
}

function _environ_get(__environ, environ_buf) {
    var bufSize = 0;
    getEnvStrings().forEach(function(string, i) {
        var ptr = environ_buf + bufSize;
        HEAP32[__environ + i * 4 >> 2] = ptr;
        writeAsciiToMemory(string, ptr);
        bufSize += string.length + 1
    });
    return 0
}

function _environ_sizes_get(penviron_count, penviron_buf_size) {
    var strings = getEnvStrings();
    HEAP32[penviron_count >> 2] = strings.length;
    var bufSize = 0;
    strings.forEach(function(string) {
        bufSize += string.length + 1
    });
    HEAP32[penviron_buf_size >> 2] = bufSize;
    return 0
}
```

that comes from `__environ.c`:

```c
#ifdef __EMSCRIPTEN__
__attribute__((constructor(100))) // construct this before user code
void __emscripten_environ_constructor(void) {
    size_t environ_count;
    size_t environ_buf_size;
    __wasi_errno_t err = __wasi_environ_sizes_get(&environ_count,
                                                  &environ_buf_size);
    if (err != __WASI_ERRNO_SUCCESS) {
        return;
    }

    __environ = malloc(sizeof(char *) * (environ_count + 1));
    if (__environ == 0) {
        return;
    }
    char *environ_buf = malloc(sizeof(char) * environ_buf_size);
    if (__environ == 0 || environ_buf == 0) {
        __environ = 0;
        return;
    }

    // Ensure null termination.
    __environ[environ_count] = 0;

    err = __wasi_environ_get((uint8_t**)__environ, environ_buf);
    if (err != __WASI_ERRNO_SUCCESS) {
        __environ = 0;
    }
}
#endif
```
This constitutes ~3KB, or about +2.3% of the total application size.

I wonder what is the machinery that is pulling this in? Iiuc LLVM treats `constructor` functions as something un-DCEable (for a reason), and `system_libs.py` has this for `libc`:
```py
    if shared.Settings.WASM_BACKEND:
      # See libc_extras below
      # Include all the getenv stuff with the wasm backend. With fastcomp we
      # still use JS because libc is a .bc file and we don't want to have a
      # global constructor there for __environ, which would mean it is always
      # included.
      libc_files += files_in_path(
          path_components=['system', 'lib', 'libc', 'musl', 'src', 'env'],
          filenames=['__environ.c', 'getenv.c', 'putenv.c', 'setenv.c', 'unsetenv.c'])
```

That looks like whenever using a) Wasm backend, and b) any symbol that will link in libc, then the 3KB of environment and the related WASI code overhead will creep in(?)

Not sure what would be the best way to avoid this, except to add a `-s POSIX_ENVIRON_SUPPORT=0` linker flag?

### PR DESCRIPTION
**Pull request by juj**

### CODE REVIEW

**Comment by sbc100** [comment]
Is there some reason why this code isn't already include on-demand?      The linker should be smart enough to exclude it when its not used, no?

**Comment by sbc100** [comment]
Another way of putting it:  Are there programs that don't depend envrion, but were paying for it code size and see and win from using SUPPORT_POSIX_ENV?      I'd been keen to see such a program to see if we can fix this in some more automatic way,  avoiding a new setting.

**Comment by juj** [comment]
I don't think so. LLVM is not able to deduce anything about functions with `constructor` attributes (how could it?). See details in #11148.

**Comment by sbc100** [comment]
Ok, thanks, I'll take a look at #11149 

The idea is that if the `constructor` lives in an object file that is not otherwise used it should not be included in the link.  In this case the `constructor` should be included if and only if some other part of the code depends on the `environ` symbol (i.e. if the program actually depends on it).

**Comment by juj** [comment]
> In this case the `constructor` should be included if and only if some other part of the code depends on the `environ` symbol

There is nothing that would make a connection between `environ` and `__emscripten_environ_constructor`. As far as I understand, `__attribute__((constructor(100)))` is a root for LLVM DCE, it cannot reason about whether that is important ctor to call or not.

The same behavior occurs btw with `emmalloc.c`s

```
static void EMSCRIPTEN_KEEPALIVE __attribute__((constructor(0))) initialize_malloc_heap()
```
If one builds with `-s MALLOC=emmalloc`, then that function will always be present, independent of whether one actually calls any malloc machinery or not.

**Comment by kripken** [comment]
> There is nothing that would make a connection between environ and __emscripten_environ_constructor. As far as I understand, __attribute__((constructor(100))) is a root for LLVM DCE, it cannot reason about whether that is important ctor to call or not.

I believe the connection is that the two are in a single C file, `__environ.c`. If and only if the linker ends up including that, will the file be linked in (because of how lld links `.a` archives - file by file, on demand).

If the file is included, then yes, the constructor will never be dce'd away. But the entire file should only be used, in this case, if you use `environ` or one of its aliases.

**Comment by juj** [comment]
If I do a debug build without optimizations enabled, I do find there is a reference to setlocale, which comes from `std::string` and `std::stringstream`. The application is using those only in a couple of places in debug builds. In release builds all the string management is optimized away by -O3/-Os DCE, and there are no std::string, std::stringstream or locale or related functions present anymore.

I wonder if what is happening is that statically first the dependency std::string->locale->getenv is present, which pulls in environ ctor, but then -O3 DCEs all of that away, but it's unable to reason about the `constructor` attribute after the fact.

One chain that I find in a debug build is

```
 (func $std::__2::num_get<char\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\20>::do_get\28std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::ios_base&\2c\20unsigned\20int&\2c\20float&\29\20const (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (param $4 i32) (param $5 i32) (result i32)
 -> calls (func $std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\20std::__2::num_get<char\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\20>::__do_get_floating_point<float>\28std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::ios_base&\2c\20unsigned\20int&\2c\20float&\29\20const (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (param $4 i32) (param $5 i32) (result i32)
    -> calls (func $std::__2::__num_get<char>::__stage2_float_prep\28std::__2::ios_base&\2c\20char*\2c\20char&\2c\20char&\29 (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (param $4 i32)
       -> calls (func $std::__2::ctype<char>::widen\28char\20const*\2c\20char\20const*\2c\20char*\29\20const (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (result i32)
         -> calls (func $__newlocale (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
             -> calls (func $__get_locale (param $0 i32) (param $1 i32) (result i32)
                 -> calls (func $getenv)
```

Removed all use of std::string and std::stringstream from the code and that causes `-s SUPPORT_POSIX_ENV=0` to no longer be needed.

In Release builds the app does not have either std::string or std::stringstream present in the .wasm, so not sure where the DCE ball dropped.

**Comment by juj** [comment]
In above, I was not able to find what is referencing
```
 (func $std::__2::num_get<char\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\20>::do_get\28std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::istreambuf_iterator<char\2c\20std::__2::char_traits<char>\20>\2c\20std::__2::ios_base&\2c\20unsigned\20int&\2c\20float&\29\20const (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (param $4 i32) (param $5 i32) (result i32)
```
but that function is only present in the debug build in function pointer table, with no direct calls to it. (In release build, that function is not present in the table)

**Comment by sbc100** [comment]
I see..   so it sounds like some kind of post-link phase is DCE'ing the getenv call but it is not possible to eliminate the static constructor.  That makes sense.   

I suspect that the elimination either occurring in binaryen or perhaps by wasm-ld when run in LTO mode.    I do not think wasm-ld would be able eliminate these calls in non-LTO mode. 

In any case, seems like a real issue, I'm just not sure how we would address it.   The static  constructor being included due to being co-located with in the same object files as a live symbol.   If some future phase decided said symbol is no longer live, I don't think we have enough information to conclude that the static constructor is can be eliminated.

**Comment by kripken** [comment]
I hope we can find a more general optimization solution here, or at least one that scales better than adding a new option + changes to `system_libs.py` for each thing like this.

Can we link in a stub `getenv()` implementation? That should prevent `__environ.c` from being linked in with that constructor. To avoid people needing to do this manually in each project, how about adding `stub_getenv.c` in emscripten somewhere, and then one would just add that file to the final link command. (We could maybe have a single new option `-s LIBC_STUBS=['getenv']` that is a shorthand for that, not sure if that is needed or not.)

**Comment by juj** [comment]
> Can we link in a stub `getenv()` implementation? That should prevent `__environ.c` from being linked in with that constructor. To avoid people needing to do this manually in each project, how about adding `stub_getenv.c` in emscripten somewhere, and then one would just add that file to the final link command. (We could maybe have a single new option `-s LIBC_STUBS=['getenv']` that is a shorthand for that, not sure if that is needed or not.)

I don't think that would work here. Spent a few hours trying to poke a minimal test case, and got one now. Updated this PR to highlight the problem in the form of a test, and showcase how `-s SUPPORT_POSIX_ENV=0` is able to optimize code size further compared to what automatic DCE at -O3 was able to. In that test, the code size shrinks from 27359 bytes to 15257 bytes (a -44% reduction) when `-s SUPPORT_POSIX_ENV=0` is used.

In my mind -s SUPPORT_POSIX_ENV is nice similar to e.g. -s SUPPORT_ERRNO, and more educating than a -s LIBC_STUBS=['how_would_i_know_what_to_add_here_to_optimize_what_and_when?']. It will also vocally complain at build time if a getenv/environ etc. is then getting referenced (and user did not want to), instead of silently generating a code size creeping stub, that one'd would then have to verify by other means it isn't present.

**Comment by kripken** [comment]
Is the testcase ` tests/env/support_posix_env.cpp`? When I build that with `emcc -O3 -o a.js` I get 19,398 bytes (13,223 js, 6,175 wasm). When I apply the idea of a stub implementation, using
```c
char **___environ = 0;
char **__environ = 0;
char **_environ = 0;
char **environ = 0;
```
then I get 11,863 bytes (11,815 js, 48 wasm - the LLVM optimizer manages to reduce main to nothing). That's a similar-sized reduction as what you see, but I'm not sure why the absolute sizes I see are smaller than yours? Did I do something wrong?

Overall I think it can be slightly unobvious how to make stubs like this (I read the musl source code to see it can define all 4 of those things), so I am in favor of adding support in emscripten for this, like a directory of stub C files. That could maybe also be usable outside of emscripten too actually.

**Comment by juj** [comment]
> I'm not sure why the absolute sizes I see are smaller than yours? Did I do something wrong?

Ops, double-checking, I had force-enabled WASM=0 as a local change applied to the tree, so the earlier numbers were for Wasm2JS. Re-doing the test on this PR branch clean, I get

```
MacBook-Pro:master clb$ ./emcc tests/env/support_posix_env.cpp -O3 -o with_support_posix_env.js -s SUPPORT_POSIX_ENV=1
MacBook-Pro:master clb$ ./emcc tests/env/support_posix_env.cpp -O3 -o with_support_posix_env_with_closure.js -s SUPPORT_POSIX_ENV=1 --closure 1
MacBook-Pro:master clb$ ./emcc tests/env/support_posix_env.cpp -O3 -o without_support_posix_env.js -s SUPPORT_POSIX_ENV=0
MacBook-Pro:master clb$ ./emcc tests/env/support_posix_env.cpp -O3 -o without_support_posix_env_with_closure.js -s SUPPORT_POSIX_ENV=0 --closure 1
ls -al with*MacBook-Pro:master clb$ ls -al with*
-rw-r--r--  1 clb  staff  13240 Jul 30 12:41 with_support_posix_env.js
-rw-r--r--  1 clb  staff   6175 Jul 30 12:41 with_support_posix_env.wasm
-rw-r--r--  1 clb  staff   6646 Jul 30 12:41 with_support_posix_env_with_closure.js
-rw-r--r--  1 clb  staff   6175 Jul 30 12:41 with_support_posix_env_with_closure.wasm
-rw-r--r--  1 clb  staff  11835 Jul 30 12:41 without_support_posix_env.js
-rw-r--r--  1 clb  staff     48 Jul 30 12:41 without_support_posix_env.wasm
-rw-r--r--  1 clb  staff   5957 Jul 30 12:41 without_support_posix_env_with_closure.js
-rw-r--r--  1 clb  staff     48 Jul 30 12:41 without_support_posix_env_with_closure.wasm
```

Without Closure: -s SUPPORT_POSIX_ENV=0 shrinks JS size from 13240 to 11835 (-10.6%), and Wasm size from 6175 to 48 (-99%) as it basically optimizes out the whole application.

Without Closure: -s SUPPORT_POSIX_ENV=0 shrinks JS size from 6646 to 5957 (-10.4%), and Wasm size the same from 6175 to 48 (-99%).

Saved the files to [artifacts.tar.gz](https://github.com/emscripten-core/emscripten/files/5000365/artifacts.tar.gz)

**Comment by kripken** [comment]
Ok good, when I compile it with closure, and with and without that
```cpp
char **___environ = 0;
char **__environ = 0;
char **_environ = 0;
char **environ = 0;
```
Then I get the same results as you (a few bytes smaller, probably noise). My previous results were without closure, which explains the last difference.

So I think this confirms that we can do this by adding another C file to the link, as opposed to adding a new setting, as I've suggested?

**Comment by juj** [comment]
> So I think this confirms that we can do this by adding another C file to the link, as opposed to adding a new setting, as I've suggested?

Ok, sounds good then. Although won't adding a new system file to the build need a new setting anyways? I wonder if a custom option will not be as discoverable?

Also, what would happen if the code does happen to utilize getenv/environ when building against the stub? (e.g. if someone accidentally adds new code to the codebase that begins to depend on getenv/environ) With the `-s SUPPORT_POSIX_ENV=0` approach the code will fail at build time, giving a clear error getenv/environ being missing. Is there a danger of getting silent failure of the code if getenv/environ are replaced with a stub?

**Comment by kripken** [comment]
I agree we should make this discoverable, and I don't have a full plan for this. It may fit into @sbc100 's ideas of a sysroot somehow, in particular? But one possibility is that this might be a single new option,

```
emcc -s STUBS=environ,foo
```

The benefit is that adding a new set of stubs is only adding a new file to some directory, and adding a test and docs, but no changes to `system_libs.py` or `settings.js` or `emcc.py` etc.

We could make the optionally stubs fail at runtime, yeah, that might be useful to guarantee things. Could be `environ_abort` in that setting sketch, for example.

**Comment by sbc100** [comment]
Should we try to break the linking between std::string -> ... -> getenv, thus allowing wasm-ld to do the DCE here?   This would at least make the specific reason for this change mute?  If we did that then even users who don't know about this option would benefit right?

I get that we probably want something with stronger guarantees though.. in care it gets included via some other path.

I'm not sure I even like `emcc -s STUBS=environ,foo` because it means more magic compilation in linking the background which I'm trying to move us away from.

I'd almost rather just say.. "include this C code in your project to avoid using our envrion code"

```
char **___environ = 0;
char **__environ = 0;
char **_environ = 0;
char **environ = 0;
```

This seems like maybe an example of the kind of place where we have traditionally made tweaks to the compiler driver, but we should be able to instead really on the input file alone.

**Comment by kripken** [comment]
> This seems like maybe an example of the kind of place where we have traditionally made tweaks to the compiler driver, but we should be able to instead really on the input file alone.

I'd like to remove compiler complexity, yes. But maybe not at all costs. Some options are:

1. Document code snippets, "compile and link this to avoid getenv() overhead".
2. Have a directory of such code snippet files, "compile and link in system/libs/stubs/getenv.c to avoid getenv() overhead".
3. Have an option to include such a code snippet file, "-s STUBS=getenv.c" (which literally just does what 2 does).

I think 1 is not friendly enough, but I think 3 adds little complexity in the compiler over 2. Maybe there's a better option though?

Btw, I think some **existing** options can be removed by turning them into such stub files, like maybe `SUPPORT_ERRNO`. It's easiest to do that if we do 3, since that's equally easy for users (replace one flag with another). That's one reason why I think this is promising.

**Comment by juj** [comment]
> We could make the optionally stubs fail at runtime, yeah, that might be useful to guarantee things.

With a `-s SUPPORT_POSIX_ENV=0` approach the failure will happen at link time, I think failing at runtime would less preferable, since it takes longer to figure out: one has to run the output and take a path at runtime where the stub getenv access occurs to find out it is being used - and if not found, the failing stub will add to code size, and become a time bomb shipped to users.

File: tests/test_other.py [hunk 1]
```diff
@@ -10481,6 +10481,27 @@ def test_native_call_after_exit(self):
...
+
+    # In optimized -O3 the test code should build properly both with and without POSIX environment support, as
+    # the POSIX environment code will be optimized away.
+    self.run_process([EMCC, path_from_root('tests', 'env', 'support_posix_env.cpp'), '-O3', '-s', 'SUPPORT_POSIX_ENV=1', '-o', 'a.html'])
+    self.run_process([EMCC, path_from_root('tests', 'env', 'support_posix_env.cpp'), '-O3', '-s', 'SUPPORT_POSIX_ENV=0', '-o', 'b.html'])
+
+    # However building with SUPPORT_POSIX_ENV=0 should optimize away environment-related code ...
+    self.assertNotContained('environ', open('b.js').read())
+    # ... and the output size should be smaller than what SUPPORT_POSIX_ENV=1 was able to do. (or otherwise -s SUPPORT_POSIX_ENV=0 is redundant option to have)
+    assert os.path.getsize('a.js') > os.path.getsize('b.js') # If this assert fails, it means -s SUPPORT_POSIX_ENV=0 is redundant, or did not work
```
Comment (sbc100) [inline tests/test_other.py:10502]: `self.assertGreater`

File: tools/system_libs.py [hunk 2]
```diff
@@ -812,6 +802,27 @@ def get_files(self):
...
+  cflags = ['-Os', '-fno-builtin']
+
+  def get_files(self):
+    if shared.Settings.WASM_BACKEND:
+      # See libc_extras below
+      # Include all the getenv stuff with the wasm backend. With fastcomp we
+      # still use JS because libc is a .bc file and we don't want to have a
+      # global constructor there for __environ, which would mean it is always
+      # included.
+      return files_in_path(
```
Comment (sbc100) [inline tools/system_libs.py:816]: Can probably drop this conditional and move the comment to the `can_build` method.

**Comment by kripken** [comment]
> I think failing at runtime would less preferable

We can easily make it fail at link time too. It could call `__getenv_is_disallowed()` which is not defined anywhere.

**Comment by juj** [comment]
Looking at the test case today, it looks like `environ` and `getenv` are correctly DCEd away from the build. So this PR is not actually needed to land. Great! Closing.

**Pull request closed without merging**